tests/map_test_case.h

```diff
@@ -0,0 +1,341 @@
+
+        /**
+        * Generates (using catch2 GENERATE) every combination of three transformations and applies them to the test.
+        * See https://github.com/catchorg/Catch2/blob/master/docs/generators.md
+        * Test will be invoked for each combination of transformation.
+        * @param t test case to apply transformations to
+        * @return string info about applied transformations
+        */
+        std::string generate_transform_combinations();
+
+        /**
+         * Ensures that `map_test_case` is set up correctly by
+         * validating given point in map coordinates against `map_test_case::anchor_map_pos`
+         * (and performing other sanity checks)
+         * @param p tripoint in map coordinates that is expected to align with the `anchor_map_pos`
+         */
+        void validate_anchor_point( tripoint p );
+
+    private:
+        // origin (0,0) of this `map_test_case` in `map` coordinates
+        // based on `anchor_map_pos` and `anchor_char`, lazily calculated when needed, reset on transformations
+        cata::optional<tripoint> origin = cata::nullopt;
+
+        // flag that internal sanity checks are completed, resets on transformations
+        bool checks_complete = false;
+
+        void do_internal_checks();
+
+        void for_each_tile( tripoint tmp_origin, const std::function<void( tile & )> &callback ) const;
+
+};
+
+// common helpers, used together with map_test_case
+namespace map_test_case_common
+{
+/**
+ * Combines two arbitrary functions `f` and `g` into a single one.
+ * `f` and `g` must have same signature
+ */
+template<typename R, typename ... Args>
+static inline std::function<R( Args ... )> operator+(
+    const std::function<R( Args ... )> f,
+    const std::function<R( Args ... )> g )
+{
+    // note: avoids std::forward for simplicity and safety
+    return [ = ]( Args ... args ) {
+        f( args ... );
+        return g( args ... );
+    };
+}
+
+/**
+ * Combines two boolean functions `f(args)` and `g(args)` into single function that returns `f(args) && g(args)`.
+ * Note, if `f` returns false, `g` is not invoked!  (&& semantics)
+ */
+template<typename ... Args>
+static inline std::function<bool( Args ... )> operator&&(
+    const std::function<bool( Args  ... )> &f,
+    const std::function<bool( Args ... )> &g
+)
+{
+    // note: avoids std::forward for simplicity and safety
+    return [ = ]( Args ... args ) {
+        return f( args... ) && g( args... );
+    };
+}
+
+/**
+ * Combines two boolean functions `f(args)` and `g(args)` into single function that returns `f(args) || g(args)`.
+ * Note, if `f` returns true, `g` is not invoked!  (|| semantics)
+ */
+template<typename ... Args>
+static inline std::function<bool( Args ... )> operator||(
+    const std::function<bool( Args  ... )> &f,
+    const std::function<bool( Args ... )> &g
+)
+{
+    return [ = ]( Args ... args ) {
+        return f( args... ) ||
+               g( std::forward<Args>( args )... );
```

Comment:
If you're going to use `std::forward` then the `Args` pack should be a pack of universal reference types; here it's a pack of value types.